modules/library/referencing/src/main/java/org/geotools/referencing/proj/PROJRefiner.java

```diff
@@ -0,0 +1,236 @@
+        }
+        for (String id : regexMap.keySet()) {
+            String regex = regexMap.get(id);
+            String replacement = replacementMap.get(id);
+            unitRefinements.add(new Refinement(regex, replacement));
+        }
+    }
+
+    private void loadRefinements() {
+        // Step 1: Identify all regex, replacement, and codes entries
+        Map<Integer, String> regexMap = new HashMap<>();
+        Map<Integer, String> replacementMap = new HashMap<>();
+        Map<Integer, String> codesMap = new HashMap<>();
+
+        for (String key : properties.stringPropertyNames()) {
+            if (key.startsWith("regex.")) {
+                regexMap.put(Integer.valueOf(key.substring(6)), properties.getProperty(key));
+            } else if (key.startsWith("replacement.")) {
+                replacementMap.put(Integer.valueOf(key.substring(12)), properties.getProperty(key));
+            } else if (key.startsWith("codes.")) {
+                codesMap.put(Integer.valueOf(key.substring(6)), properties.getProperty(key));
+            }
+        }
+
+        // Step 2: Process each regex code and associate with EPSG codes
+        List<Integer> sortedKeys = new ArrayList<>(regexMap.keySet());
+        Collections.sort(sortedKeys);
+        for (int id : sortedKeys) {
+            String regex = regexMap.get(id);
+            String replacement = replacementMap.get(id);
+            String codes = codesMap.get(id);
+
+            if (regex != null && replacement != null && codes != null) {
+                List<String> epsgCodes = parseCodes(codes);
+                for (String code : epsgCodes) {
+                    epsgRefinements
+                            .computeIfAbsent(code, k -> new ArrayList<>())
+                            .add(new Refinement(regex, replacement));
+                }
+            }
+        }
+    }
+
+    private List<String> parseCodes(String codes) {
+        List<String> result = new ArrayList<>();
+        String[] codeParts = codes.split(",");
+        for (String part : codeParts) {
+            if (part.contains("-")) {
+                String[] range = part.split("-");
+                int start = Integer.parseInt(range[0]);
+                int end = Integer.parseInt(range[1]);
+                for (int i = start; i <= end; i++) {
+                    result.add(String.valueOf(i));
+                }
+            } else {
+                result.add(part);
+            }
+        }
+        return result;
+    }
+
+    // Apply global updates and regex replacements for specific EPSG codes
+    public String refine(String projString, String epsgCode) {
+        // Apply global updates
+        StringBuilder updatedProjString = new StringBuilder(sortProjString(projString));
+        updatedProjString.append(" " + globalAdditions);
+
+        // First refine the units
+        for (Refinement refinement : unitRefinements) {
+            Pattern pattern = Pattern.compile(refinement.regex);
```

Comment:
This is bad... compiling patterns is expensive, and the patterns we have here are static once the application is started (they won't be reloaded/changed). So please parse the regular expressions on load, rather than re-compiling them all for each and every proj string GeoTools is generating.